### PR TITLE
feat(api): brand untrusted prompt input with sanitizeForPrompt

### PR DESCRIPTION
**Pull request by jan-kubica**
## Summary

Adds branded types and a sanitiser that make it harder to accidentally interpolate external content into an LLM prompt without going through a visible promotion step.

- `UntrustedText` — declares a string crossed an untrusted boundary (uploaded document, user paste, external API response, other tenants' data).
- `PromptSafeText` — only mintable via `sanitizeForPrompt`.
- `sanitizeForPrompt(text, opts?)` strips ASCII control chars (preserving tab/LF/CR), Unicode bidi and zero-width overrides used in published injection PoCs, ChatML / Llama / generic role-prefix markers, and the delimiter sequences themselves so input cannot pre-close the wrapper. Output is wrapped in distinctive `<<<UNTRUSTED>>>…<<<END_UNTRUSTED>>>` markers the surrounding template can refer to as a data boundary.

This is defence in depth, not a guarantee. It pairs with least-privilege tool use, output filtering, and tenant partitioning. What the brand does guarantee: a contributor cannot interpolate raw input into a prompt position without going through a function named `sanitizeForPrompt`, visible in code review.

## Scope

Intentionally minimal: ships the library + tests only. Adoption at existing call sites (e.g. `chat-prompt.ts`) and a lint rule that forbids raw strings flowing into prompt builders are follow-ups so each can be reviewed on its own merits.

## Test plan

- [x] `bun test apps/api/src/lib/prompt-safety.test.ts` — 12 pass / 0 fail
- [x] `bun run lint` — clean across all 22 packages
- [x] `bun run typecheck` — clean across all 22 packages

### CODE REVIEW

File: apps/api/src/lib/prompt-safety.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  let cleaned: string = stripControlChars(text);
+  cleaned = cleaned.replace(INVISIBLE_OVERRIDES, "");
+  for (const pattern of ROLE_MARKERS) {
+    cleaned = cleaned.replace(pattern, "");
+  }
+  cleaned = cleaned.split(open).join("").split(close).join("");
+
+  if (options?.maxLength !== undefined && cleaned.length > options.maxLength) {
+    cleaned = `${cleaned.slice(0, options.maxLength)}${TRUNCATION_SUFFIX}`;
+  }
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/prompt-safety.ts:165]: Fixed: truncation now avoids cutting between UTF-16 surrogate pairs, with a regression test for the boundary case.

CC on behalf of @jan-kubica

File: apps/api/src/lib/prompt-safety.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+const stripControlChars = (input: string): string => {
+  let out = "";
+  for (const ch of input) {
+    const code = ch.codePointAt(0);
+    if (code === undefined || !isStrippableControl(code)) {
+      out += ch;
+    }
+  }
+  return out;
+};
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/prompt-safety.ts]: Addressed with the follow-up sanitizer change: the control-character stripper now builds chunks and joins once, preserving linear behavior. I kept the explicit scanner instead of a control-character regex because the repo lint policy rejects control regex literals.

CC on behalf of @jan-kubica

[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/prompt-safety.ts]: Fixed: stripControlChars no longer concatenates on every removed character; it records clean slices and joins once, so the pass is O(n).

CC on behalf of @jan-kubica

File: apps/api/src/lib/prompt-safety.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+): PromptSafeText => {
+  const open = options?.open ?? DEFAULT_OPEN;
+  const close = options?.close ?? DEFAULT_CLOSE;
+
+  let cleaned: string = stripControlChars(text);
+  cleaned = cleaned.replace(INVISIBLE_OVERRIDES, "");
+  for (const pattern of ROLE_MARKERS) {
+    cleaned = cleaned.replace(pattern, "");
+  }
+  cleaned = cleaned.split(open).join("").split(close).join("");
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/prompt-safety.ts]: Fixed: delimiter removal now replaces each full delimiter with a space instead of joining around it, so split delimiter fragments cannot reassemble into a live delimiter.

CC on behalf of @jan-kubica

File: apps/api/src/lib/prompt-safety.ts [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ * (used in published injection PoCs), ChatML / Llama / generic
+ * role-prefix lines, and the delimiter sequences themselves (so
+ * input cannot pre-close the wrapper).
+ */
+export const sanitizeForPrompt = (
+  text: UntrustedText,
+  options?: SanitizeForPromptOptions,
+): PromptSafeText => {
+  const open = options?.open ?? DEFAULT_OPEN;
+  const close = options?.close ?? DEFAULT_CLOSE;
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/prompt-safety.ts]: Addressed: sanitizeForPrompt now rejects blank custom open/close delimiters instead of returning PromptSafeText without a real fence.

CC on behalf of @jan-kubica

File: apps/api/src/lib/prompt-safety.ts [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+    if (chunkStart < index) {
+      chunks.push(input.slice(chunkStart, index));
+    }
+    chunkStart = index + 1;
+  }
+
+  if (chunks.length === 0) {
+    return input;
+  }
```
[the remark this answers is not in the document]
Comment (jan-kubica) [reply apps/api/src/lib/prompt-safety.ts]: Addressed: stripControlChars now tracks whether any control character was found, so leading-only and all-control inputs are stripped correctly; regression tests cover both cases.

CC on behalf of @jan-kubica
